### ISSUE TITLE
No way to register a class method as callback in the same class?

### ISSUE THREAD

**Issue by lilydjwg** [vim/vim#12417]
### Steps to reproduce

Define a class within whose constructor method (`new`), try to register a method as callback:

1.
```vim
vim9script

class MyClass
  this.job: job

  def new()
    this.job = job_start('cat', {
      out_cb: this.OnRequest, # <--- here
    })
  enddef

  def OnRequest(ch: channel, msg: string)
    echo msg
  enddef

endclass
```

=> `E722: Missing comma in Dictionary: OnRequest,`

2. If changed to `out_cb: function('this.OnRequest'),` => `E475: Invalid argument: this.OnRequest`

3.
```vim
vim9script

def OnRequest(c: MyClass, ch: channel, msg: string)
  echo msg
  c.Do()
enddef

class MyClass
  this.job: job

  def new()
    this.job = job_start('cat', {
      out_cb: function(OnRequest, [this]),
    })
  enddef

  def Do()
  enddef

endclass

var t = MyClass.new()
```

=> `E1010: Type not recognized: MyClass`

4.
```vim
vim9script

def OnRequest(c: any, ch: channel, msg: string)
  echo msg
  c.Do()
enddef

class MyClass
  this.job: job

  def new()
    this.job = job_start('echo a', {
      out_cb: function(OnRequest, [this]),
    })
  enddef

  def Do()
  enddef

endclass

var t = MyClass.new()
```
=>
```
Error detected while processing function <SNR>185_OnRequest:
line    2:
E715: Dictionary required 
```

5.
```vim
vim9script

def OnRequest(c: any, ch: channel, msg: string)
  echo msg
  c.Do()
enddef

class MyClass
  this.job: job

  def new()
    this.job = job_start('echo a', {
      out_cb: function(OnRequest, [], this),
    })
  enddef

  def Do()
  enddef

endclass

var t = MyClass.new()
```
=>
```
Error detected while compiling /home/lilydjwg/tmpfs/a.vim[22]..function new:
line    2:
E1013: Argument 3: type mismatch, expected dict<any> but got object<MyClass> 
```

6. 

### Expected behaviour

There should be a way to register class method as callback inside the class' method.

### Version of Vim

9.0.1566

### Environment

Arch Linux

**Comment by brammool** [comment]
Please try 9.0.1569 or later.  At least the first example appears to be fixed now.

**Comment by lilydjwg** [comment]
With 9.0.1570, the same error is only delayed to the time when the class is initialized, e.g.

```vim
vim9script

class MyClass
  this.job: job

  def new()
    this.job = job_start('cat', {
      out_cb: this.OnRequest, # <--- here
    })
  enddef

  def OnRequest(ch: channel, msg: string)
    echo msg
  enddef

endclass

var c = MyClass.new()
```

And it's back.

**Comment by errael** [comment]
A workaround with a bit of indirection
```
vim9script

class MyClass
  this.job: job

  def new()
    var CB = (ch: channel, msg: string) => this.OnRequest(ch, msg)
    this.job = job_start('cat', {
      out_cb: CB, # <--- here
    })
  enddef

  def OnRequest(ch: channel, msg: string)
    echo msg
  enddef

endclass

var c = MyClass.new()
```

**Comment by lilydjwg** [comment]
Thanks, that worked!

**Comment by chrisbra** [comment]
closed by 29bb67f1beefc7fd393dbfd9ee77d92f1db3a3c0
